middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,101 @@
+import socket
+
+from typing import Iterator
+from typing import Optional
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+from middleman_protocol.utils import append_frame_separator
+from middleman_protocol.utils import escape_decode_raw_message
+from middleman_protocol.utils import escape_encode_raw_message
+
+
+def split_stream(connection: socket.socket) -> Iterator[bytes]:
+    """
+    Lowest level receiver which determines if received data is frame separator,
+    which means previously received data should be yielded,
+    or if received data is something else and receiver should continue to gather bytes.
+    """
+    assert isinstance(connection, socket.socket)
+
+    received_data = b''
+
+    while True:
+        next_bytes = connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        for next_byte in next_bytes:
+            if next_bytes == FRAME_SEPARATOR:
+                yield received_data
+                received_data = b''
+            else:
+                received_data += bytes([next_byte])
+
+
+def unescape_stream(connection: socket.socket) -> Iterator[Optional[bytes]]:
+    """
+    Middle level receiver which determines if received data has broken escaping and yields None to indicate it,
+    or reverses escaping and yields unescaped frame.
+    """
+    assert isinstance(connection, socket.socket)
+
+    for unescaped_data in split_stream(connection=connection):
+        assert isinstance(unescaped_data, bytes)
+
+        # If there is an escape character which is not a valid escape sequence,
+        # yield special value indicating a frame with broken escaping and continue the outer loop.
+        # This can be checked by comparing sum of occurrences of escape character
+        # to sum of occurrences of escape sequences. If they are equal, data is escaped correctly.
+        if (
+            unescaped_data.count(ESCAPE_CHARACTER) !=
+            sum([unescaped_data.count(escape_sequence) for escape_sequence in ESCAPE_SEQUENCES.values()])
+        ):
```

Comment:
I don't really understand this one. Can you elaborate ?

`escape_decode_raw_message` should be simple utility function, the reverse of `escape_encode_raw_message`.

What do you mean by ` it won't detect invalid sequences.` ?

Reply:
> `escape_decode_raw_message` should be simple utility function, the reverse of `escape_encode_raw_message`.

But it's not the reverse as defined now. `escape_decode_raw_message()` accepts invalid input that can never be produced by `escape_encode_raw_message()`.

> What do you mean by ` it won't detect invalid sequences.` ? 

I mean that if `escape_decode_raw_message()` gets `XXX\aYYY` as input and `\` is the escape character but `\a` is not defined as a valid sequence then that function will include `\a` in its output instead of telling you that the input is invalid.

This is inconsistent with the generator - which does detect invalid input.

I think that the check should be inside `escape_decode_raw_message()`, not inside the generator so that both work the same way for invalid input.

Reply:
Ok, now I understand, changed and added test. :+1: done